File: src/main/java/testla/screenplay/actor/Actor.java
package testla.screenplay.actor;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import testla.screenplay.ability.IAbility;
import testla.screenplay.action.IAction;
import testla.screenplay.question.IQuestion;
import testla.screenplay.task.ITask;

/**
 * Actors use abilities in order to execute tasks/actions and answer questions.
 *
 * @author Patrick Doering
 */
public class Actor implements IActor {

    // collection of attributes assigned to the actor
    Map<String, Object> attributes = new HashMap<>();

    // map of abilities of this Actor
    private final Map<String, IAbility> abilityMap = new HashMap<>();

    private Actor(String name) {
        this.attributes.put("name", name);
    }

    /**
     * Create a new Actor with a given name.
     *
     * @param name the name of the new actor.
     * @return the newly created Actor.
     */
    public static Actor named(String name) {
        return new Actor(name);
    }

    /**
     * Store an attribute in the actors attribute collection.
     *
     * @param key the attribute name.
     * @param value the attribute value.
     * @return itself.
     */
    @Override
    public Actor with(String key, Object value) {
        this.attributes.put(key, value);
        return this;
    }

    /**
     * Get an attribute from the actors attribute collection. If no attribute is stored under the given key, returns null.
     *
     * @param key the key for the attribute.
     * @return the value for the requested attribute if it exists, null if not.
     */
    @Override
    public Object states(String key) {
        return this.attributes.get(key);
    }

    /**
     * Ask a question.
     *
     * @param question the question to ask.
     * @return the question result.
     */
    @Override
    public <T> T asks(IQuestion<T> question) {
        return question.answeredBy(this);
    }

    /**
     * Verify if the actor has the given ability.
     *
     * @param ability the ability.
     * @return the associated ability if the actor has it.
     */
    @Override
    public IAbility withAbilityTo(IAbility ability) {
        if (ability == null || !this.abilityMap.containsKey(ability.name())) {
            throw new RuntimeException("Error: This Actor does not have this ability!");
        }
        return this.abilityMap.get(ability.name());
    }

    /**
     * Assign one or more abilities to the actor.
     *
     * @param abilities the abilities the actor will be able to use.
     * @return itself.
     */
    @Override
    public Actor can(IAbility... abilities) {
        Arrays.stream(abilities).forEach(ability -> this.abilityMap.put(ability.name(), ability));
        return this;
    }

    /**
     * Executes the given Tasks.
     *
     * @param tasks a list of tasks to execute.
     * @return the return value of the last task executed.
     */
    @Override
    public Object attemptsTo(ITask... tasks) {
        Object[] taskResult = new Object[1];
        Arrays.stream(tasks).forEach(task -> taskResult[0] = task.performAs(this));
        return taskResult[0];
    }

    /**
     * Executes the given Actions.
     *
     * @param actions a list of actions to execute.
     * @return the return value of the last action executed.
     */
    @Override
    public Object attemptsTo(IAction... actions) {
        Object[] actionResult = new Object[1];
        Arrays.stream(actions).forEach(action -> actionResult[0] = action.performAs(this));
        return actionResult[0];
    }
}

File: src/main/kotlin/de/p7s1/testla/screenplay/Actor.kt
package de.p7s1.testla.screenplay
import de.p7s1.testla.*

class Actor private constructor(name: String) : IActor {
    // map of attributes that are stored to the actor
    override var attributes: MutableMap<String, Any> = mutableMapOf()

    // map of abilities of this Actor
    private var abilityMap: MutableMap<String, IAbility> = mutableMapOf()

    // constructor continued
    init {
        this.attributes["name"] = name
    }

    // static methods here
    companion object {
        /**
         * Create a new Actor with a given name.
         * @param name the actor's name
         */
        @JvmStatic
        fun named(name: String): Actor {
            return Actor(name)
        }
    }

    /**
     * Store an attribute in the actors attribute collection.
     * @param key attribute name
     * @param value attribute value
     * @returns this Actor
     */
    override fun with(key: String, value: Any): IActor {
        this.attributes[key] = value
        return this
    }

    /**
     * Get an attribute from the actors attribute collection.
     * @param key Key for the attribute
     * @returns Value for the requested attribute
     */
    override fun states(key: String): Any? {
        return this.attributes[key]
    }

    /**
     * Sets username and password for the actor.
     *
     * @param username
     * @param password
     * @returns the actor object
     *
     * @deprecated This method is deprecated and will be removed in the future. Use
     */
    fun withCredentials(username: String, password: String): Actor {
        this.attributes["username"] = username
        this.attributes["password"] = password
        return this
    }

    override fun <T> asks(question: IQuestion<T>, activityResult: T): T {
        return question.answeredBy(this)
    }

    override fun withAbilityTo(ability: IAbility): IAbility {
        if (!abilityMap.containsKey(ability.name)) {
            throw Error("Error: This Actor does not have this ability! ({${ability.name})")
        } else {
            return abilityMap[ability.name]!!
        }
    }

    override fun can(vararg abilities: IAbility): IActor {
        abilities.forEach {
            this.abilityMap[it.name] = it
        }
        return this
    }

    /**
     * Executes the given Tasks/Actions.
     *
     * @param activities a list of tasks to execute.
     */
    override fun attemptsTo(vararg activities: IAction): Any {
        // execute each activity in order.
        val result = activities.forEach {
            it.performAs(this)
        }
        return result
    }
}